examples/pipeline/wav2letter.py

```diff
@@ -0,0 +1,1179 @@
+        self.iterable = iterable
+        self._iter = iter(iterable)
+        self._done = False
+        self._values = []
+
+    def __iter__(self):
+        if self._done:
+            return iter(self._values)
+        return itertools.chain(self._values, self._gen_iter())
+
+    def _gen_iter(self):
+        for new_value in self._iter:
+            self._values.append(new_value)
+            yield new_value
+        self._done = True
+
+    def __len__(self):
+        return len(self._iterable)
+
+
+class MapMemoryCache(torch.utils.data.Dataset):
+    """
+    Wrap a dataset so that, whenever a new item is returned, it is saved to memory.
+    """
+
+    def __init__(self, dataset):
+        self.dataset = dataset
+        self._cache = [None] * len(dataset)
+
+    def __getitem__(self, n):
+        if self._cache[n]:
+            return self._cache[n]
+
+        item = self.dataset[n]
+        self._cache[n] = item
+
+        return item
+
+    def __len__(self):
+        return len(self.dataset)
+
+
+class Processed(torch.utils.data.Dataset):
+
+    def __init__(self, process_datapoint, dataset):
+        self.process_datapoint = process_datapoint
+        self.dataset = dataset
+
+    def __getitem__(self, n):
+        try:
+            item = self.dataset[n]
+            return self.process_datapoint(item)
+        except (FileNotFoundError, RuntimeError):
+            return None
+
+    def __next__(self):
+        try:
+            item = next(self.dataset)
+            return self.process_datapoint(item)
+        except (FileNotFoundError, RuntimeError):
+            return self.__next__()
+
+    def __len__(self):
+        return len(self.dataset)
+
+
+def process_datapoint(item):
+    transformed = item[0]  # .to(device, non_blocking=non_blocking)
+    target = item[2].lower()
+
+    transformed = transforms(transformed)
+
+    transformed = transformed[0, ...].transpose(0, -1)
+
+    target = " " + target + " "
+    target = encode(target)
+    target = torch.tensor(target, dtype=torch.long, device=transformed.device)
+
+    transformed = transformed  # .to("cpu")
+    target = target  # .to("cpu")
+    return transformed, target
+
+
+def datasets_librispeech():
+
+    def create(tag):
+
+        if isinstance(tag, str):
+            data = LIBRISPEECH(
+                root, tag, folder_in_archive=folder_in_archive, download=False)
+        else:
+            data = torch.utils.data.ConcatDataset([LIBRISPEECH(
+                root, t, folder_in_archive=folder_in_archive, download=False) for t in tag])
+
+        data = Processed(process_datapoint, data)
+        # data = diskcache_iterator(data)
+        data = MapMemoryCache(data)
+        return data
+
+    return create("train-clean-100"), create("dev-clean"), None
+    # return create(["train-clean-100", "train-clean-360", "train-other-500"]), create(["dev-clean", "dev-other"]), None
+
+
+def which_set(filename, validation_percentage, testing_percentage):
```

Comment:
Why is this necessary as opposed to a seeded shuffle + split?

Reply:
This is the script recommended for splitting train/dev/test in SpeechCommands' readme adapted to this use case. I suggest we include it with SpeechCommands in torchaudio.

Reply:
How are they better distributed in comparison to a random shuffle?

Reply:
Sorry, the real advantage of this approach is listed in the docstring:

> We want to keep files in the same training, validation, or testing sets even if new ones are added over time. This makes it less likely that testing samples will accidentally be reused in training when long runs are restarted for example.